packages/coinjoin/src/client/round/outputDecomposition.ts

```diff
@@ -167,20 +60,325 @@ export const getOutputAmounts = async (params: GetOutputAmountsParams) => {
+    log(`Change amount credentials: ${change.amountCredentials.map(c => c.value).join(',')}`);
+    log(`Change vsize credentials: ${change.vsizeCredentials.map(c => c.value).join(',')}`);
+
+    return {
+        output,
+        change,
+    };
+};
+
+const findCredentialsForTarget = (
+    target: number,
+    credentials: middleware.Credentials[],
+    maxValue: number,
+): [middleware.Credentials, middleware.Credentials] | undefined => {
+    // sort descending. higher possibility for match
+    const sorted = credentials.sort((a, b) => (a.value > b.value ? -1 : 1));
+
+    // find one Credential big enough to cover requested target
+    const bigCredential = sorted.find(cre => cre.value >= target);
+    if (bigCredential) {
+        // try find a pair to produce change Credential used in next iteration
+        // always try to pair with greatest Credential or at least 0 value Credential
+        const changeValue = bigCredential.value - target;
+        const pair = sorted.find(
+            cre => cre !== bigCredential && cre.value + changeValue <= maxValue,
+        );
+        if (pair) {
+            return [bigCredential, pair];
+        }
+        // this should never happen but just in case
+        throw new Error('Missing pair for credential');
+    }
+
+    // one Credential is not enough. try to find pair to cover requested target
+    const candidate = sorted
+        .map(cre => {
+            // always try to pair with greatest Credential to produce greatest possible change
+            const pair = sorted.find(
+                p => p !== cre && p.value + cre.value >= target && p.value + cre.value <= maxValue,
+            );
+            if (pair) {
+                return [cre, pair];
+            }
+            return [];
+        })
+        .find(pair => pair.length === 2);
+
+    return candidate ? [candidate[0], candidate[1]] : undefined;
+};
+
+interface Bob {
+    accountKey: string;
+    amount: number;
+    amountCredentials: middleware.Credentials[];
+    vsizeCredentials: middleware.Credentials[];
+}
+
+interface CreateOutputsCredentials {
+    round: CoinjoinRound;
+    options: CoinjoinRoundOptions;
     accountKey: string;
     outputSize: number;
     amounts: number[];
     amountCredentials: middleware.Credentials[];
     vsizeCredentials: middleware.Credentials[];
+    result: Bob[];
+}
+
+const createOutputsCredentials = async (params: CreateOutputsCredentials): Promise<Bob[]> => {
+    const { round, amounts, outputSize, amountCredentials, vsizeCredentials, options } = params;
+    const { roundParameters } = round;
+
+    // sort amounts ascending. smaller amounts should produce bigger change Credentials for next iteration
+    const sorted = amounts.sort((a, b) => a - b);
+
+    // try to find Credentials for each amount
+    const amountsWithCredentials = sorted.map(amount => ({
+        amount,
+        credentials: findCredentialsForTarget(
```

Comment:
If I understand it correctly, I think that `findCredentialsForTarget` could be called only once for `amountPair` and once for `vsizePair` and moreover, without the amount params. If it just returns the biggest possible pair not exceeding the limit, we could then decide here whether it's enough for the smallest amount or we have to join instead. Am I right?

Reply:
it's an array of { amount, credentials? }:
1. try to find Credentials for each amount
1. then try to find **first** possible pair (amountPair declaration)
1. if not exists execute [code below](https://github.com/trezor/trezor-suite/pull/7206#discussion_r1049801967) (condition if (amountPair?.credentials ...)

Reply:
I see now. I was calculating the limit in my head incorrectly. Resolved.